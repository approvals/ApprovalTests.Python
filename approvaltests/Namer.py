import inspect
import os


class Namer(object):
    ClassName = ''
    MethodName = ''
    Directory = ''

    APPROVED = '.approved'
    RECEIVED = '.received'

    def __init__(self, frame=1, extension='.txt'):
        self.extension_with_dot = extension
        self.frame = frame
        self.set_for_stack(inspect.stack(1))

    def get_class_name(self):
        return self.ClassName

    def get_method_name(self):
        return self.MethodName

    def get_directory(self):
        return self.Directory

    def get_basename(self):
        return os.path.join(self.Directory, self.ClassName + "." + self.MethodName)

    def get_received_filename(self, basename=None):
        basename = basename or self.get_basename()
        return basename + self.RECEIVED + self.extension_with_dot

    def get_approved_filename(self, basename=None):
        basename = basename or self.get_basename()
        return basename + self.APPROVED + self.extension_with_dot

    def set_for_stack(self, caller):
        stacktrace = caller[self.frame]
        self.MethodName = stacktrace[3]
        self.ClassName = stacktrace[0].f_locals["self"].__class__.__name__
        self.Directory = os.path.dirname(stacktrace[1])
