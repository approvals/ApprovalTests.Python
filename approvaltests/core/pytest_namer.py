
class PytestNamer:
    """
    For use with parameterized pytest tests.

    Use this namer when the same test case needs to verify more than one value, and produce more than one file.
    """
    def __init__(self, base_namer, pytest_request):
        self.base_namer = base_namer
        if pytest_request.node.originalname:
            self.parameters = pytest_request.node.name[len(pytest_request.node.originalname):]
        else:
            self.parameters = ''

    def get_basename(self):
        basename = self.base_namer.get_basename()
        return basename + self.parameters

    def get_approved_filename(self, basename=None):
        if basename is None:
            basename = self.get_basename()
        return self.base_namer.get_approved_filename(basename)

    def get_received_filename(self, basename=None):
        if basename is None:
            basename = self.get_basename()
        return self.base_namer.get_received_filename(basename)
