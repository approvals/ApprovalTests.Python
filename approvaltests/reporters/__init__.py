"""Approvaltests.reporters module."""
from .clipboard_reporter import *
from .default_reporter_factory import *
from .diff_reporter import *
from .first_working_reporter import *
from .generic_diff_reporter import *
from .generic_diff_reporter_config import *
from .generic_diff_reporter_factory import *
from .multi_reporter import *
from .python_native_reporter import *
from .received_file_launcher_reporter import *
from .reporter_that_automatically_approves import *
from .report_all_to_clipboard import *
from .report_by_creating_diff_file import *
