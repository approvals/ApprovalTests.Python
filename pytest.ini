[pytest]
testpaths = tests
addopts = --showlocals

asyncio_default_fixture_loop_scope = function
