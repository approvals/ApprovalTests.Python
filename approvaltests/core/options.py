from typing import Callable, Dict, Optional

from approvaltests.core.comparator import Comparator
from approvaltests.core.namer import Namer
from approvaltests.core.reporter import Reporter
from approvaltests.file_approver import FileComparator
from approvaltests.inline.inline_options import InlineOptions
from approvaltests.scrubbers import combine_scrubbers


class FileOptions:
    def __init__(self, fields: Dict):
        self.fields = fields

    @property
    def file_extention(self) -> str:
        return self.fields.get("extension_with_dot", ".txt")

    def with_extension(
        self,
        extension_with_dot: str,
        *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/,
        no_override: bool = False,
    ) -> "Options":
        if not extension_with_dot.startswith("."):
            extension_with_dot = "." + extension_with_dot
        if no_override and "extension_with_dot" in self.fields:
            extension_with_dot = self.fields["extension_with_dot"]
        return Options({**self.fields, "extension_with_dot": extension_with_dot})


class Options:
    def __init__(self, fields: Optional[Dict] = None):
        self.fields = fields or {}

    @property
    def reporter(self) -> Reporter:
        from approvaltests.reporters.default_reporter_factory import (
            get_default_reporter,
        )

        return self.fields.get("reporter", get_default_reporter())

    @property
    def comparator(self) -> Comparator:
        return self.fields.get("comparator", FileComparator())

    def with_comparator(self, comparator: Comparator) -> "Options":
        return Options({**self.fields, "comparator": comparator})

    def scrub(self, data: str) -> str:
        if self.has_scrubber():
            return self.fields["scrubber_func"](data)
        return data

    def with_scrubber(self, scrubber_func: Callable[[str], str]) -> "Options":
        return Options({**self.fields, "scrubber_func": scrubber_func})

    def add_scrubber(self, scrubber: Callable[[str], str]) -> "Options":
        if self.has_scrubber():
            scrubber = combine_scrubbers(self.fields["scrubber_func"], scrubber)
        return self.with_scrubber(scrubber)

    def has_scrubber(self) -> bool:
        return "scrubber_func" in self.fields

    def with_reporter(self, reporter: "Reporter") -> "Options":
        return Options({**self.fields, "reporter": reporter})

    def with_namer(self, namer: "Namer") -> "Options":
        return Options({**self.fields, "namer": namer})

    @property
    def for_file(self) -> FileOptions:
        return FileOptions(self.fields)

    @property
    def namer(self) -> Namer:
        from approvaltests.namer.default_name import get_default_namer

        namer = self.fields.get("namer", get_default_namer())
        if hasattr(namer, "set_extension"):
            namer.set_extension(self.for_file.file_extention)
        return namer

    def inline(self, inline_options: Optional[InlineOptions] = None) -> "Options":
        from approvaltests.namer.inline_comparator import InlineComparator

        return InlineComparator().register(self, inline_options)

    def add_reporter(self, additional_reporter: Reporter) -> "Options":
        from approvaltests.reporters import MultiReporter

        return self.with_reporter(MultiReporter(self.reporter, additional_reporter))
