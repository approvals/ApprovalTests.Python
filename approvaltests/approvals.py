from threading import local
import xml.dom.minidom

from approvaltests import to_json
from approvaltests.approval_exception import ApprovalException
from approvaltests.core.namer import Namer
from approvaltests.core.scenario_namer import ScenarioNamer
from approvaltests.core.pytest_namer import PytestNamer
from approvaltests.file_approver import FileApprover
from approvaltests.list_utils import format_list
from approvaltests.reporters.diff_reporter import DiffReporter
from approvaltests.string_writer import StringWriter

DEFAULT_REPORTER = local()


def set_default_reporter(reporter):
    global DEFAULT_REPORTER
    DEFAULT_REPORTER.v = reporter
    

def get_default_reporter():
    if not hasattr(DEFAULT_REPORTER, "v") or DEFAULT_REPORTER.v is None:
        return DiffReporter()
    return DEFAULT_REPORTER.v


def get_reporter(reporter):
    return reporter or get_default_reporter()


def get_default_namer():
    return Namer()


def verify(data, reporter=None, namer=None):
    reporter_to_use = reporter or get_default_reporter()
    namer_to_use = namer or get_default_namer()
    verify_with_namer(data, namer_to_use, reporter_to_use)


def verify_with_namer(data, namer, reporter):
    reporter = get_reporter(reporter)
    approver = FileApprover()
    writer = StringWriter(data)
    error = approver.verify(namer, writer, reporter)
    if error is not None:
        raise ApprovalException(error)


def verify_as_json(object, reporter=None):
    n_ = to_json(object) + "\n"
    verify(n_, reporter)


def verify_xml(xml_string, reporter=None):
    try:
        dom = xml.dom.minidom.parseString(xml_string)
        pretty_xml = dom.toprettyxml()
    except Exception:
        pretty_xml = xml_string
    verify_with_namer(pretty_xml, Namer(extension='.xml'), reporter)


def verify_file(file_name, reporter=None):
    with open(file_name, 'r') as f:
        file_contents = f.read()
    verify(file_contents, reporter)


def verify_all(header, alist, formatter=None, reporter=None):
    text = format_list(alist, formatter, header)
    verify(text, reporter)


def get_scenario_namer(scenario_name):
    return ScenarioNamer(get_default_namer(), scenario_name)

def get_pytest_namer(pytest_request):
    return PytestNamer(get_default_namer(), pytest_request)