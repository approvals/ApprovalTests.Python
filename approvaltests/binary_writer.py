from typing import ByteString

from typing_extensions import override

from approval_utilities.utils import create_directory_if_needed
from approvaltests.core.writer import Writer


class BinaryWriter(Writer):
    contents: ByteString

    def __init__(
        self,
        contents: ByteString,
        extension: str,
    ) -> None:
        self.contents = contents
        self.extension_with_dot = extension

    @override
    def write_received_file(self, received_file: str) -> str:
        create_directory_if_needed(received_file)
        with open(received_file, mode="wb") as file:
            file.write(self.contents)

        return received_file
