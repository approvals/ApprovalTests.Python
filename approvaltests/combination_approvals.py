from itertools import product
from typing import Any, Callable, List, Optional, Sequence

from approval_utilities.utilities.logger.simple_logger import SimpleLogger
from approvaltests.approvals import initialize_options, verify
from approvaltests.core.options import Options
from approvaltests.core.reporter import Reporter
from approvaltests.pairwise_combinations import get_best_covering_pairs
from approvaltests.reporters.testing_reporter import ReporterForTesting

VariationForEachParameter = Sequence[Sequence[Any]]
CombinationsOfParameters = Sequence[Sequence[Any]]


def calculate_total_size(input_arguments: VariationForEachParameter) -> int:
    from functools import reduce

    return reduce(
        lambda current_size, arguments: len(arguments) * current_size,
        input_arguments,
        1,
    )


def verify_best_covering_pairs(
    function_under_test: Callable,
    input_arguments: VariationForEachParameter,
    formatter: Optional[Callable] = None,
    reporter: Optional[ReporterForTesting] = None,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
) -> None:
    combinations = get_best_covering_pairs(input_arguments)
    count = len(combinations)

    text = print_combinations(formatter, function_under_test, combinations)
    options = initialize_options(options, reporter)
    total = calculate_total_size(input_arguments)

    header = f"Testing an optimized {count}/{total} scenarios:\n\n"
    verify(header + text, options=options)


def run_all_combinations(
    function_under_test: Callable,
    input_arguments: VariationForEachParameter,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    exception_handler: Callable[[BaseException], None] = SimpleLogger.warning,
) -> None:
    parameter_combinations = product(*input_arguments)
    for args in parameter_combinations:
        try:
            function_under_test(*args)
        except BaseException as exception:
            exception_handler(exception)


def verify_all_combinations_with_labeled_input(
    function_under_test: Callable,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
    **kwargs: Any,
) -> None:
    labels = list(kwargs.keys())
    input_arguments = [kwargs[key] for key in kwargs]

    def formatter(inputs: Sequence[Any], output: Any) -> str:
        labeled_inputs = ", ".join(
            [f"{label}: {input}" for label, input in zip(labels, inputs)]
        )
        return f"({labeled_inputs}) => {output}\n"

    verify_all_combinations(
        function_under_test, input_arguments, formatter=formatter, options=options
    )


def verify_all_combinations(
    function_under_test: Callable,
    input_arguments: VariationForEachParameter,
    formatter: Optional[Callable] = None,
    reporter: Optional[ReporterForTesting] = None,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
) -> None:
    """Run func with all possible combinations of args and verify outputs against the recorded approval file.

    Args:
        function_under_test (function): function under test.
        input_arguments: list of values to test for each input argument.  For example, a function f(product, quantity)
            could be tested with the input_arguments [['water', 'cola'], [1, 4]], which would result in outputs for the
            following calls being recorded and verified: f('water', 1), f('water', 4), f('cola', 1), f('cola', 4).
        formatter (function): function for formatting the function inputs/outputs before they are recorded to an
            approval file for comparison.
        reporter (approvaltests.reporter.Reporter): an approval reporter.

    Raises:
        ApprovalException: if the results to not match the approved results.
    """
    options = initialize_options(options, reporter)
    verify_all_combinations_with_namer(
        function_under_test, input_arguments, formatter, None, options=options
    )


def verify_all_combinations_with_namer(
    function_under_test: Callable,
    input_arguments: VariationForEachParameter,
    formatter: Optional[Callable] = None,
    reporter: Optional[Reporter] = None,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
) -> None:
    """Run func with all possible combinations of args and verify outputs against the recorded approval file.

    Args:
        function_under_test (function): function under test.
        input_arguments: list of values to test for each input argument.  For example, a function f(product, quantity)
            could be tested with the input_arguments [['water', 'cola'], [1, 4]], which would result in outputs for the
            following calls being recorded and verified: f('water', 1), f('water', 4), f('cola', 1), f('cola', 4).
        namer (approvaltests.core.Namer): A namer that defines the name of received and approved files.
        formatter (function): function for formatting the function inputs/outputs before they are recorded to an
            approval file for comparison.
        reporter (approvaltests.reporter.Reporter): an approval reporter.

    Raises:
        ApprovalException: if the results to not match the approved results.
    """
    text = print_combinations(formatter, function_under_test, product(*input_arguments))
    options = initialize_options(options, reporter)
    verify(text, options=options)


def print_combinations(
    formatter: Optional[Callable],
    function_under_test: Callable,
    parameter_combinations: CombinationsOfParameters,
) -> str:
    if formatter is None:
        formatter = args_and_result_formatter
    approval_strings = []
    for args in parameter_combinations:
        try:
            result = function_under_test(*args)
        except BaseException as exception:
            result = exception
        approval_strings.append(formatter(args, result))
    return "".join(approval_strings)


def args_and_result_formatter(args: List[Any], result: int) -> str:
    return f"args: {repr(args)} => {repr(result)}\n"


def verify_logging_for_all_combinations(
    function_to_run: Callable,
    input_arguments: VariationForEachParameter,
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
) -> None:
    def printer(*args: Any) -> None:
        SimpleLogger._wrapper.get().log_line(f"Running inputs {args} => ")
        with SimpleLogger._wrapper.get().indent():
            function_to_run(*args)

    output = SimpleLogger.log_to_string()
    run_all_combinations(printer, input_arguments)
    verify(output, options=options)
