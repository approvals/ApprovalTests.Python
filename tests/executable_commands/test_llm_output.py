from approvaltests import verify, verify_executable_command
import pytest

from approval_utilities.approvaltests.core.executable_command import ExecutableCommand

from openai import OpenAI

client = OpenAI()


def call_llm(prompt: str) -> str:
    pass


class CallLlm(ExecutableCommand):
    def get_command(self) -> str:
        return """
        please rename the name foo to smthg better

        ---

        ```
        function foo(a,b) {
        return a +b
        }
        ```

        output format:
        suggested name - <name here>
        """

    def execute_command(self, command: str) -> str:
        completion = client.chat.completions.create(
            model="gpt-4-1106-preview",
            messages=[
                {"role": "system", "content": "You are a helpful assistant."},
                {"role": "user", "content": command},
            ],
        )
        return completion.choices[0].message.content


def test_total_output():
    verify_executable_command(CallLlm())


def get_new_name(llm_output):
    pass


@pytest.mark.skip()
def test_useful_output():
    verify(get_new_name(llm_output=reuse_recorded_llm_output()))
