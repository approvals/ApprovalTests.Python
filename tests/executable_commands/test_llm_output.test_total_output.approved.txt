
        please rename the name foo to smthg better

        ---

        ```
        function foo(a,b) {
        return a +b
        }
        ```

        output format:
        suggested name - <name here>
        
